### PR TITLE
add dynamodb database access

### PR DESCRIPTION
**Pull request by GavinFrazar**
This PR adds DynamoDB support to database access #17842

DAX, dynamodb, and dynamodb streams are all supported.

## Why
We want to bring DynamoDB into database access for consistency and to enable some things not possible via app access.
This change will enable the use of NoSQLWorkbench GUI client for DynamoDB.
 
## How
Configure one Teleport database for each AWS region you want in your cluster.
  - This is what enables NoSQL Workbench via "local" connections, because otherwise it sends requests to the non-existent "local" AWS region, and I cannot find a way to get it to respect HTTPS_PROXY env variable.

## Example Configuration
```yaml
  databases:
    - name: "home-dynamodb"
      protocol: "dynamodb"
# optional uri, if uri is set then AWS region can be extracted from that
# or if AWS region is already set then the regions must match.
      # uri: "dynamodb.us-west-2.amazonaws.com:443"
      static_labels:
        env: "dev"
      aws:
        region: "us-west-2"
        account_id: "278576220453"
```

## RBAC
* `--db-name` is not supported.
* `--db-user` should be an AWS role, either "role/SomeRole" or "SomeRole" is fine. The full ARN is constructed using configured AWS account ID and the AWS partition (determined by region).

## tsh
* DynamoDB has no "shell-session" type of cli that I can find, which is unique among the db protocols we support. Therefore `tsh db connect` prints a user message telling them to use `tsh proxy db --tunnel` instead.
* "The connect command" in `tsh db ls -v` is instead `tsh proxy db --tunnel` as well.
* For now we require the `--tunnel` flag. I may add non-tunnel later, but it will only work with the aws cli, not NoSQLWorkbench, since NoSQLWorkbench does not have a way to configure trusted CA bundle.
 
### Example usage:
1. Setup local proxy tunnel: `$ tsh proxy db --tunnel --port 8000 --db-user=GavinDynamoDBRole home-dynamodb`
2. (optionally) use aws cli: `$ aws dynamodb list-tables --endpoint-url=http://localhost:8000`
3. (optionally) use NoSQLWorkbench:
* click "Launch" Amazon DynamoDB > "Operation builder" > "Add connection" > "DynamoDB local" tab.
* Name the connection as you want and match the local proxy port setting, then click "connect".
* All operations thereon will be in the configured AWS region.
image
4. (optionally) use AWS sdk for your apps, for example python w/ boto3:
```python3
$ python3
Python 3.10.4 (main, Mar 31 2022, 03:37:37) [Clang 12.0.0 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import boto3
>>> clt = boto3.client('dynamodb', endpoint_url='http://localhost:8000')
>>> res = clt.list_tables()
>>> print(res)
{'TableNames': *snip output*}
>>> 
```
This PR is stacked against the app access PR #19387 for merge.

TODO: update webapps repo to recognize the DynamoDB db access request event codes.

### CODE REVIEW

**Comment by GavinFrazar** [comment]
@smallinsky this is ready I just need to finish the writeup before I mark it ready for review by others, but since you are already familiar with what this PR accomplishes you can take a look if you want

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:152]: whoops, I was debugging something and forgot to change that back. yes it should be as you suggested

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:146]: thanks for catching, I agree and generally prefer to avoid named return. These places are just remnants of me putting in `defer .. if err != nil` debug statements that I missed when cleaning up my changes

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:81]: My understanding is that each new connection dispatches an entirely new db engine instance. This is the first engine that I've added, but I see other engines doing the same things with the conn, context, etc, without locks.

File: api/types/database.go
Comment (GavinFrazar) [reply api/types/database.go:695]: I can remove it. I was not aware that we prefer to avoid logging in `api`, since we do it in `DatabaseV3.CheckAndSetDefaults` for other databases. Why do we prefer to avoid logrus in api?

File: api/go.mod
Comment (GavinFrazar) [reply api/go.mod:6]: This came from `api/utils/aws/endpoint.go`, specifically the `ParseDynamoDBServiceFromTarget` function, where I just use 3 constants from the sdk. I can hardcode those as they are unlikely to change, or I can just move that function out of api since it is only referenced by test code which is not necessary.

File: api/types/database.go
Comment (GavinFrazar) [reply api/types/database.go:695]: ok hmm. I'd like to log this message. If I remove it now with plans to "eventually" add it back when #12786 is addressed to inject a logger, then it will probably just be forgotten... I don't like adding to the problem, but I'm not sure how else I could log the message without substantial rework

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:81]: There would be a race, but we don't call SendError in a separate goroutine.

https://github.com/gravitational/teleport/blob/e5fca6e1be07d68a7d98b9c1cadcb3f945d074e5/lib/srv/db/server.go#L772-L780

File: api/go.mod
Comment (GavinFrazar) [reply api/go.mod]: I moved functions that didn't need to be in `api` into the engine itself in `lib` along with the tests for those funcs, removed the `aws-sdk-go` dependency in `api`, and added tests in `api/utils/aws/endpoint_test.go` for `DynamoDBRegionForEndpoint`, because I had apparently forgotten to add a test for that 😬

File: lib/srv/db/dynamodb/engine.go
Comment (smallinsky) [inline lib/srv/db/dynamodb/engine.go:107]: Could you add a reference to this format ?

File: lib/srv/db/dynamodb/engine.go
Comment (smallinsky) [inline lib/srv/db/dynamodb/engine.go:103]:
```suggestion
// jsonErr is used to marshal a JSON error response as the AWS CLI expects for errors.
```

File: lib/srv/db/dynamodb/engine.go
Comment (smallinsky) [inline lib/srv/db/dynamodb/engine.go:138]:
```suggestion
		Body: io.NopCloser(bytes.NewBuffer(body)),
```

File: lib/srv/db/dynamodb/engine.go
Comment (smallinsky) [inline lib/srv/db/dynamodb/engine.go:179]:
```suggestion
	serviceID, err := e.getServiceID(req)
```

File: api/types/database.go
Comment (smallinsky) [inline api/types/database.go:498]: Can we rename the `checkAndSetDynamoDBDefaults` to something like `handleDynamoDBConfig`  `process...`.  The `checkAndSet` default functions mostly are called at the beginning of function and  should not contain any complex "buisness logic".

File: lib/srv/db/dynamodb/engine.go
Comment (smallinsky) [inline lib/srv/db/dynamodb/engine.go:60]: Nit: I think that we recently updated the logic how db engines are registered. 
https://github.com/gravitational/teleport/blob/master/lib/srv/db/server.go#L59

File: lib/srv/db/dynamodb/engine.go
Comment (greedy52) [inline lib/srv/db/dynamodb/engine.go:362]: nit: do we need this loop?

File: lib/srv/db/dynamodb/engine.go
Comment (greedy52) [inline lib/srv/db/dynamodb/engine.go:263]: 👍

File: lib/srv/db/dynamodb/engine.go
Comment (greedy52) [inline lib/srv/db/dynamodb/engine.go:220]: Do we log an audit event when the request fails anywhere? Feels we should capture at least what the client sends.

File: lib/srv/db/dynamodb/engine.go
Comment (greedy52) [inline lib/srv/db/dynamodb/engine.go:379]: 😄 AWS really not consistent eh

File: lib/srv/db/dynamodb/engine.go
Comment (greedy52) [inline lib/srv/db/dynamodb/engine.go:341]: nit: personally i would use [endpoints.DefaultResolver().EndpointFor](https://docs.aws.amazon.com/sdk-for-go/api/aws/endpoints/#Partition.EndpointFor) to create the url using the `ServiceID` and `Region`

File: tool/tsh/db.go
Comment (greedy52) [inline tool/tsh/db.go:293]:
```suggestion
```

File: tool/tsh/proxy.go
Comment (greedy52) [inline tool/tsh/proxy.go:873]: I think you can curl ElastiCache in HTTPS without `--tunnel`. @Tener to confirm.

I assume non-tunnel mode works for DynamoDB as well as long the client uses the right cert and CA? I am ok though if we want to keep it simple by only supporting tunnel mode.

**Review by greedy52 — COMMENTED** [review]
First pass!

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:107]: reference: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html#Programming.Errors.Components

Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:107]: although I found this originally by just digging deep into the aws sdk to see how they unmarshal errors. Before I added this the `aws` cli would print a useless generic error like "failed to unmarshal error response..." because it would try and fail to read the response error as json

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:60]: will fix when I merge master in

File: api/types/database.go
Comment (GavinFrazar) [reply api/types/database.go:678]: I have been conflicted about this as well. One thing I've realized is that fips endpoints will not work currently, since they will get rewritten to the standard endpoint like `dynamodb-fips.us-west-2.amazonaws.com` -> `dynamodb.us-west-2.amazonaws.com`. At the same time, we determine the endpoint based on the incoming request at run time... I may need to rework the configuration logic such that the following behavior holds:
1. blank uri is derived from the region and set to the standard endpoint suffix.
2. custom uri is allowed and we just try to derive region from the uri if we need to and can do so.
3. don't override any custom uri. But in the engine, check to see if the configured uri is an aws endpoint missing the service subdomain and add the subdomain if that is the case.

In this configuration logic, if the user sets the uri like:
`URI: "dynamodb.us-west-2.amazonaws.com"` then the service subdomain will not be overwritten and any requests sent to that uri which are trying to do streams operations or DAX will just fail. This is something we can document in the guide, as in "don't specify the URI if you want to allow dax/streams/dynamodb requests to be proxied by the agent. If you want fips, then you need to specify the fips uri".

Comment (GavinFrazar) [reply api/types/database.go:678]: The alternative to this would be to force users to setup a static database for each service endpoint and for each region they want. I think that's not very friendly UX, but I'm open to doing that if others think that is better

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:136]: They do recognize the format, I've tested with aws cli and NoSQL Workbench. I considered using the amz content type, but this error isn't actually coming from amazon so I opted for the standard content type header for json.

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:209]: we do not.

@smallinsky we discussed adding this before and iirc decided to not add external id support in the initial implementation, unless I'm mistaken, but to be sure: should we support this? If we do then I think putting it in `db_service->databases[i].aws.external_id` would be the appropriate place.

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:220]: I didn't consider that other engines emit the audit event regardless of roundtrip error

File: tool/tsh/proxy.go
Comment (GavinFrazar) [reply tool/tsh/proxy.go:873]: Did you mean ElasticSearch?

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:379]: lol

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:362]: actually I think we don't, thanks for point that out!

File: api/types/database.go
Comment (greedy52) [reply api/types/database.go:678]: > 1. blank uri is derived from the region and set to the standard endpoint suffix.
> 2. custom uri is allowed and we just try to derive region from the uri if we need to and can do so.
> 3. don't override any custom uri. But in the engine, check to see if the configured uri is an aws endpoint missing the service subdomain and add the subdomain if that is the case.

I think this is reasonable.  `region` with blank URI is likely the "default" way ppl would use most of the time. And it's fine to ask users to specify a custom URI to overwrite it when the "default" way doesn't work. 

The only part I find a little confusing personally is having the `.us-west-2.amazonaws.com` as the URI (you would also see this in `tsh db ls -v` too). One other alternative is to use a custom schema when there is a blank URI (for example, populate the URI to something like `aws://dynamodb`. Krzysztof did something similar for the Azure CLI support for app access. This way, you can tell the difference between a blank URI vs a custom URI, and if need to, force the user to specify a full/valid URI when not doing the "default" way.

> The alternative to this would be to force users to setup a static database for each service endpoint and for each region they want.

I think this is also fine since this is the case for other databases today. Personally, I do like the above-quoted logic better.

File: lib/srv/db/dynamodb/engine.go
Comment (greedy52) [reply lib/srv/db/dynamodb/engine.go:136]: nice!

File: lib/srv/db/dynamodb/engine.go
Comment (greedy52) [reply lib/srv/db/dynamodb/engine.go:220]: most database engines capture the client requests for the audit and do not wait for the response. so I guess we care more about what the client sends than what the server responds.

File: api/utils/aws/endpoint.go
Comment (greedy52) [reply api/utils/aws/endpoint.go:698]: yes, you have to manually specify the URI for cassandra-fips.

File: lib/srv/db/dynamodb/engine.go
Comment (smallinsky) [reply lib/srv/db/dynamodb/engine.go:209]: The sounds ok and handling `external_id`  looks like minor change so we can add it now. 
> If we do then I think putting it in db_service->databases[i].aws.external_id would be the appropriate place.

Yea, that we can att  it will be consistent with current app access approach where external_id is in handled by `apps[i].aws.external_id`

File: api/types/database.go
Comment (GavinFrazar) [reply api/types/database.go:695]: Changed the config check/set function so this is obsolete. I do not log anything anymore because I don't override a custom URI.

File: api/types/database.go
Comment (GavinFrazar) [reply api/types/database.go:678]: Ok I've updated the uri logic as described in my initial suggestion above.

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:220]: Updated to log an audit event for the request when roundtrip fails to get a response

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go:341]: refactored endpoint resolution around this concept. It simplified things greatly!

File: tool/tsh/proxy.go
Comment (GavinFrazar) [reply tool/tsh/proxy.go:873]: leaving this as-is until @Tener confirms, but I believe that *if* a local proxy is used with elastic search, then it requires the --tunnel flag. See: https://github.com/gravitational/teleport/blob/7aa224e430f708dfb53f365f32871a1e0ea01010/tool/tsh/db.go#L577-L582

Comment (greedy52) [reply tool/tsh/proxy.go:873]: `tsh db connect` uses `elasticsearch-sql-cli` which supports tunnel mode only. `tsh proxy aws` calls dbcmd's `GetConnectCommandAlternatives()` on the other hand.

**Review by greedy52 — APPROVED** [review]
Looks good to me. Also tested it out today and worked great!

One note I believe `dax` support we have here today is only for for dax APIs. 

I doubt if there are any customer needs [dax endpoints](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DAX.concepts.cluster.html#DAX.concepts.cluster-endpoint) access from Teleport (as we can access dynamodb directly). If so, we may consider doing dax endpoints support separately where use AWS API for discovery and [dax client](https://github.com/aws/aws-dax-go/) for engine. I can create a ticket for this.

File: api/types/database.go
Comment (greedy52) [reply api/types/database.go:678]: Cool. Do we still need this `if awsutils.IsAWSEndpoint(d.Spec.URI)` check?

File: lib/events/codes.go
Comment (GavinFrazar) [reply lib/events/codes.go:193]: thank you. I could've sworn I changed that suffix to "E" 😞

File: tool/tsh/db.go
Comment (GavinFrazar) [reply tool/tsh/db.go:1115]: Sounds good to me, I'll try refactoring to be similar to the login templating logic which prints connect/proxy/config commands conditionally using those `formatDatabase*` helpers

File: tool/tsh/proxy.go
Comment (Tener) [reply tool/tsh/proxy.go:873]: > I think you can curl Elasticsearch in HTTPS without `--tunnel`. @Tener to confirm.

This is right. For Elasticsearch `--tunnel` is simply much easier for clients. Making curl use custom certs is easy; the same thing is much harder with db clients.

Comment (GavinFrazar) [reply tool/tsh/proxy.go:873]: fixed

File: tool/tsh/db.go
Comment (GavinFrazar) [reply tool/tsh/db.go:1115]:
```
[I] [14:01:59] gavin@mac ~/teleport-config [1] 
$ tsh db env mysql                                                                                                                     (base) 
ERROR: "tsh db env" is not supported for MySQL databases when TLS routing is enabled on the Teleport Proxy Service.

Please use one of the following commands to connect to the database:
    `tsh db connect --cluster=mars.local.gd --db-user=<user> mysql`
    `tsh proxy db --tunnel --cluster=mars.local.gd --db-user=<user> mysql`

[I] [14:02:02] gavin@mac ~/teleport-config [1] 
$ tsh db config cass
ERROR: "tsh db config" is not supported for Cassandra databases.

Please use one of the following commands to connect to the database:
    `tsh db connect --cluster=mars.local.gd --db-user=<user> cass`
    `tsh proxy db --tunnel --cluster=mars.local.gd --db-user=<user> cass`

[I] [14:02:06] gavin@mac ~/teleport-config [1] 
$ tsh db login --db-user=GavinDynamoDBRole dynamodb
Connection information for database "dynamodb" has been saved.

You can start a local proxy for database GUI clients:

  tsh proxy db --tunnel --cluster=mars.local.gd dynamodb

[I] [14:02:36] gavin@mac ~/teleport-config  
$ tsh db connect dynamodb
ERROR: "tsh db connect" is not supported for DynamoDB databases.

Please use the following command to connect to the database:
    `tsh proxy db --tunnel --cluster=mars.local.gd dynamodb`
[I] [14:02:40] gavin@mac ~/teleport-config [1] 
$ tsh proxy db dynamodb 
ERROR: "tsh proxy db" is not supported for DynamoDB databases without the --tunnel flag.

Please use the following command to connect to the database:
    `tsh proxy db --tunnel --cluster=mars.local.gd dynamodb`
```

How's this?

Comment (greedy52) [reply tool/tsh/db.go]: I am not very good at UX either. Does `--cluster` flag always show up or only when `--cluster` is passed in? And personally I would also prefer without the back quotes so it's easier to copy/paste

Comment (GavinFrazar) [reply tool/tsh/db.go]: Oh, it shows the cluster because I had the env var `TELEPORT_CLUSTER` set. Without that it doesn't show the cluster flag.

I'll remove the backticks though

Comment (GavinFrazar) [reply tool/tsh/db.go]: `--db-user=` flag only shows up because I didn't specify it when I did `tsh db login`, else that will also not appear in the suggested command.

**Review by smallinsky — APPROVED** [review]

File: lib/srv/db/dynamodb/engine.go
Comment (GavinFrazar) [reply lib/srv/db/dynamodb/engine.go]: refactored this

File: tool/tsh/db.go
Comment (greedy52) [reply tool/tsh/db.go]: Sounds good!

**Review by greedy52 — APPROVED** [review]

**Comment by GavinFrazar** [comment]
Due to PR size this requires admin approval to merge - who can do that? @r0mant @smallinsky

**Comment by smallinsky** [comment]
> Due to PR size this requires admin approval to merge - who can do that?
I think that we need ask @r0mant or @zmb3  for approval.

**Review by r0mant — APPROVED** [review]
Bot.

**Pull request merged**
